### PR TITLE
Alternative implementation for wxGenericValidator

### PR DESCRIPTION
**Pull request by AliKet**
This will replace PR #831 

### Motivations:

The new approach tries to overcome the flawed implementation of `wxGenericValidator::Transfer[To|From]Window()`
(i.e. keep on doing a bunch of `wxDynamicCast`'s and testings whenever we want to do any transfer of data is not ideal)

furthermore, what prevents us from doing something like this:

```
std::optional<wxString> opstr;
win->SetValidator(wxGenericValidator(opstr));
```
...if `win` knows how to transfer a variable of type `wxString` ?

---------------------------------

### Usage:

With the newly introduced `wxSetGenericValidator()` function template we can do:

```
int m_sel;
bool m_ok;
wxString m_str;
wxScopedPtr<wxString> m_strPtr;
.
.
auto ctrl_1 = new wxRadioBox(...);
auto ctrl_2 = new wxComboBox(...);

wxSetGenericValidator(ctrl_1, &m_str); // Error: wxRadioBox does not support wxString's
wxSetGenericValidator(ctrl_1, &m_ok); // Error: wxRadioBox does not support bool's
wxSetGenericValidator(ctrl_1, &m_sel); // Fine
wxSetGenericValidator(ctrl_2, m_strPtr); // Note the absence of the ampersand op in front of m_strPtr
wxSetGenericValidator(ctrl_2, &m_strPtr); // Error: wxComboBox does not support wxScopedPtr<wxString>
```

**Bonus:**

if a window `win` does support multiple types (e.g. `int` and `wxString` in the case of `wxComboBox`) 
you may choose to use `std::variant` as follows:

```
std::variant<int, wxString> m_data; // make sure that all the types are supported by ctrl
std::variant<int, bool, wxString> m_data2;
auto ctrl_1 = new wxComboBox(...);
auto ctrl_2 = new wxComboBox(...);

wxSetGenericValidator(ctrl_1, m_data); // Note the absence of the ampersand op in front of m_data
wxSetGenericValidator(ctrl_2, m_data2); // this won't compile as wxComboBox doesn't support bool data type

```

### CODE REVIEW

**Comment by AliKet** [comment]
Hi @vadz 

Please feel free to review this at your leisure, TIA!

**Comment by AliKet** [comment]
Hi @vadz and sorry for taking so long to reply

Concerning your first remark about transferring data from a combobox to  `variant<int, string>` , your're right, `variant` is not appropriate here, and something like tuple would be the proper type i think.
anyway, i moved this specialization to the validate sample, just in case anyone needs it in the future.

About your second remark, i've implemented this with wxMonoValidatorPanel<> which is essentially a wxPanel with any number of controls supporting data transfer, but only one of them is the active control (i.e. the last one that has its value changed) which will go through data validation and eventually get its data transferred.

**Comment by vadz** [comment]
Sorry for a very long delay, unfortunately I don't really expect to be able to address this before 3.2.0 any longer. But, just in case I do have some time to look at it, is this one still relevant or was it replaced by #2380?

**Comment by AliKet** [comment]
> Sorry for a very long delay, unfortunately I don't really expect to be able to address this before 3.2.0 any longer. But, just in case I do have some time to look at it, is this one still relevant or was it replaced by #2380?

Closing because this PR is a mess (as it is now) and is hard to follow/review (and sorry for that)
but if ever we need to do anything to improve wxGenericValidator, then we need a clean/clear strategy to do that.

sorry and thank you for bringing this up.
